sender.go

```diff
@@ -160,13 +160,23 @@ func (s *Sender) send(ctx context.Context, msg *Message, opts *SendOptions) (cha
 			fr.Done = make(chan encoding.DeliveryState, 1)
 		}
 
+		// NOTE: we MUST send a copy of fr here since we modify it post send
+
+		sent := make(chan error, 1)
 		select {
-		case s.transfers <- fr:
+		case s.transfers <- transferEnvelope{Ctx: ctx, Frame: fr, Sent: sent}:
 			// frame was sent to our mux
 		case <-s.l.done:
 			return nil, s.l.doneErr
-		case <-ctx.Done():
```

Comment:
Removing this is responsible for the test failure and raises an interesting question.

For the affected test, the sender is attempting to send a message when there are no credits on the link.  Adding back the `<-ctx.Done()` returns us to the ambiguous state.  And while it's highly unlikely that a sending link would never have any credit, letting the call hang if the context has a deadline/is cancelable is counter intuitive.

Given our previous goal of stating that a canceled/expired context only happens when waiting for an ack, I think the only thing we can do here is terminate the sender and return a `*LinkError`.  I think this is reasonable considering that if a sender doesn't have link credit, it's a symptom of a larger problem.  What do you think?

Reply:
This seems right to me. 

If we flip it around, we're basically talking about a case where the Receiver has issues a FLOW request, we exhausted all the credits and then we decide, hey, we're going to send you some bonus messages. That's a constraint violation to me - the Receiver isn't expecting to receive any data so we shouldn't be sending any.

Reply:
Also, I don't think it's unlikely to have zero link credit - this must happen all the time from Service Bus's perspective after all credits have been used.

Reply:
Per discussion, this will return a transfer limit exceeded error now.